File: back/src/data/transactions.ts
import { tz } from "@date-fns/tz";
import { eachMonthOfInterval } from "date-fns";

import type { Pg } from "./data.ts";
import { id } from "./id.ts";

export function transactions(sql: Pg) {
	return {
		stats: async ({
			userId,
			start,
			end,
			timezone,
		}: {
			userId: string;
			start: Date;
			end: Date;
			timezone: string;
		}) => {
			const rows = await sql`
select
	t.id as id,
	t.date as date,
	t.amount as amount,
	t.category_id as category_id,
	c.name as cat_name,
	c.is_neutral as cat_is_ne,

	link.id as link_id, 
	link.transaction_a_id as link_transaction_a_id, 
	link.transaction_b_id as link_transaction_b_id,

	linked.id as linked_id,
	linked.amount as linked_amount
from transactions t
left join transactions_links link on link.transaction_a_id = t.id or link.transaction_b_id = t.id
left join transactions linked on (
	link.transaction_b_id = linked.id and link.transaction_a_id = t.id
) or (
	link.transaction_a_id = linked.id and link.transaction_b_id = t.id
)
left join transaction_categories c on t.category_id = c.id
where t.user_id = ${userId}
and t.date at time zone ${timezone} between ${start} and ${end}
and c.is_neutral = false;
`;

			const transactions = {} as {
				[key: string]: {
					id: string;
					period: string;
					amount: number;
					category: {
						id: string;
						name: string;
						is_neutral: boolean;
					} | null;
					links: Array<{
						id: string;
						transaction: {
							id: string;
							amount: number;
						};
					}>;
				};
			};

			// map rows
			for (let i = 0; i < rows.length; i++) {
				const row = rows[i];

				const id = row.id as string;
				const link_id = row.link_id as string;

				let t = transactions[id];

				if (!t) {
					t = {
						id,
						period: (row.date as Date).toISOString().slice(0, 7),
						amount: row.amount as number,
						category: null,
						links: [],
					};

					const categoryId = row.category_id as string | undefined;
					if (categoryId && !t.category) {
						const category_name = row.cat_name as string;
						const is_neutral = row.cat_is_ne as boolean;
						t.category = {
							id: categoryId,
							name: category_name,
							is_neutral,
						};
					}
				}

				if (link_id) {
					const linked_id = row.linked_id as string;
					const hasLink = t.links.find(
						(l) => l.transaction.id === linked_id || l.id === link_id
					);
					if (hasLink) continue;

					const amount = row.linked_amount as number;

					t.links.push({
						id: link_id,
						transaction: {
							id: linked_id,
							amount,
						},
					});
				}

				transactions[id] = t;
			}

			// apply linked transactions
			for (const transaction_id in transactions) {
				const t = transactions[transaction_id];

				if (
					// skip 0 amount transaction, already processed
					t.amount === 0 ||
					// skip neutral categories
					// TODO: maybe show these somehow?
					t.category?.is_neutral
				) {
					continue;
				}

				// negate linked transactions accordingly
				if (t.links.length) {
					if (t.amount > 0) {
						// positive tx
						// go through each negative link "paying"
						// for the linked tx with the positive tx
						// effectively reducing positive tx amount
						// and increasing linked tx amount
						for (let i = 0; i < t.links.length; i++) {
							const link = t.links[i];
							// interested in negative links only
							if (link.transaction.amount > 0) continue;

							const linkedTx = transactions[link.transaction.id];

							const newAmount = Math.max(t.amount + linkedTx.amount, 0);
							const newLinkedAmount = Math.min(linkedTx.amount + t.amount, 0);
							linkedTx.amount = newLinkedAmount;
							link.transaction.amount = newLinkedAmount;

							t.amount = newAmount;
						}

						// above loop is done, `t.amount` is up-to-date
						// update that amount to linked.links
						for (let i = 0; i < t.links.length; i++) {
							const link = t.links[i];
							// interested in negative links only
							if (link.transaction.amount > 0) continue;

							const tx = transactions[link.transaction.id];
							const thisTxLinkIdx = tx.links.findIndex(
								(l) => l.transaction.id == t.id
							);
							tx.links[thisTxLinkIdx].transaction.amount = t.amount;
						}
					}
				}
			}

			const timeframe = eachMonthOfInterval({ start, end }, { in: tz(timezone) });

			const neg_categories = new Set<string>();
			const pos_categories = new Set<string>();

			let total_neg = 0;
			let total_pos = 0;

			type YYYYMM = string;
			type ChartData = {
				__period__: YYYYMM;
				__total_pos__: number;
				__total_neg__: number;
			};
			const period_totals: Record<
				YYYYMM,
				ChartData | ({ [key: string]: number } & ChartData)
			> = {};

			for (const t_id in transactions) {
				const t = transactions[t_id];
				if (t.amount === 0) continue;
				const category_name = t.category?.name ?? "__uncategorized__";

				if (!period_totals[t.period]) {
					period_totals[t.period] = {
						__period__: t.period,
						__total_neg__: 0,
						__total_pos__: 0,
					};
				}

				// @ts-expect-error -- index signature etc
				period_totals[t.period][category_name] =
					// @ts-expect-error -- index signature etc
					(period_totals?.[t.period]?.[category_name] ?? 0) + t.amount;

				if (t.amount > 0) {
					period_totals[t.period].__total_pos__ += t.amount;
					pos_categories.add(category_name);
					total_pos += t.amount;
				} else {
					period_totals[t.period].__total_neg__ += t.amount;
					neg_categories.add(category_name);
					total_neg += t.amount;
				}
			}

			const chart_data = new Array(timeframe.length);

			for (let i = 0; i < timeframe.length; i++) {
				const p = timeframe[i].toISOString().slice(0, 7);
				chart_data[i] = period_totals[p] ?? {
					__period__: p,
					__total_neg__: 0,
					__total_pos__: 0,
				};
			}

			return {
				totalNeg: total_neg,
				totalPos: total_pos,
				posCategories: [...pos_categories.values()],
				negCategories: [...neg_categories.values()],
				stats: chart_data,
			};
		},

		statsCumulative: async ({
			userId,
			timezone,
			start,
			end,
		}: {
			userId: string;
			timezone: string;
			start: Date;
			end: Date;
		}) => {
			return await sql`
with months as (
	select generate_series(
		date_trunc('month', ${start} at time zone ${timezone}), 
		date_trunc('month', ${end} at time zone ${timezone}),
		'1 month'
	) as month
),
monthly_balances as (
	select 
		date_trunc('month', date at time zone ${timezone}) as month, 
		sum(amount) as monthly_sum
	from transactions
	where user_id = ${userId} and date between ${start} and ${end}
	group by month
),
cumulative as (
	select 
		m.month as date, 
		coalesce(sum(mb.monthly_sum) over (order by m.month rows between unbounded preceding and current row), 0) as value
	from months m
	left join monthly_balances mb on m.month = mb.month
)
select date, value from cumulative order by date;
			`;
		},

		query: async (opts: {
			userId: string;
			cursor?: { id: string; dir: "left" | "right" };
			limit?: number;
			query?: string;
		}) => {
			const { userId, query } = opts;

			let cursorClause = null;
			let order = sql`desc`;
			const limit = (opts?.limit ?? 50) + 1;
			const cursor = opts?.cursor;

			if (cursor) {
				const { id, dir } = cursor;

				if (dir === "left") {
					order = sql`asc`;
					cursorClause = sql`
					(
						(
							t.date = (select date from transactions where id = ${id})
							and t.id > ${id}
						)
						or t.date > (select date from transactions where id = ${id})
					)`;
				} else if (dir === "right") {
					cursorClause = sql`
					(
						(
							t.date = (select date from transactions where id = ${id})
							and t.id < ${id}
						)
						or t.date < (select date from transactions where id = ${id})
					)`;
				}
			}

			const queryClause = query
				? sql`and t.ts @@ plainto_tsquery('english', ${query})`
				: null;

			const rows = await sql`
				select
					t.id                  as tx_id,
					t.date                as tx_date,
					t.amount	          as tx_amount,
					t.currency            as tx_currency,
					t.additional          as tx_additional,
					t.counter_party       as tx_counter_party,
					t.category_id         as category_id,

					c.name       as cat_name,
					c.is_neutral as cat_is_neutral,

					linked.id as linked_id,
					linked.date as linked_date,
					linked.amount as linked_amount,
					linked.currency as linked_currency,
					linked.additional as linked_additional,
					linked.counter_party as linked_counter_party,

					link.id as link_id,
					link.created_at as link_created_at,
					link.updated_at as link_updated_at

				from transactions t

				left join transaction_categories c
					on t.category_id = c.id

				left join transactions_links link
					on link.transaction_a_id = t.id or link.transaction_b_id = t.id
				left join transactions linked
					on link.transaction_b_id = linked.id and link.transaction_a_id = t.id 
					or link.transaction_a_id = linked.id and link.transaction_b_id = t.id

				where t.user_id = ${userId}
				${queryClause ? sql`${queryClause}` : sql``}
				${cursorClause ? sql`and ${cursorClause}` : sql``}
				order by t.date ${order}, t.id ${order}
				limit ${limit};
			`;

			const hasMore = rows.length === limit;
			if (hasMore) {
				rows.pop();
			}

			if (cursor?.dir === "left") {
				rows.reverse();
			}

			const transactions = new Map<string, TransactionWithLinks>();

			for (let i = 0; i < rows.length; i++) {
				const row = rows[i];

				const id = row.tx_id as string;
				const linkedTransactionId = row.linked_id as string;

				let t = transactions.get(id);

				if (!t) {
					t = {
						id,
						date: row.tx_date as Date,
						amount: row.tx_amount as number,
						currency: row.tx_currency as string,
						additional: row.tx_additional as string,
						counter_party: row.tx_counter_party as string,
						category: null,
						links: [],
					};

					const categoryId = row.category_id as string | undefined;
					if (categoryId) {
						const categoryName = row.cat_name as string;
						const is_neutral = row.cat_is_neutral as boolean;
						t.category = {
							id: categoryId,
							name: categoryName,
							is_neutral,
						};
					}
				}

				if (linkedTransactionId) {
					const lDate = row.linked_date as Date;
					const lAmount = row.linked_amount as number;
					const lCurrency = row.linked_currency as string;
					const lAdditional = row.linked_additional as string;
					const lCounterParty = row.linked_counter_party as string;
					const linkCreatedAt = row.link_created_at as Date;
					const linkUpdatedAt = row.link_updated_at as Date;
					const linkId = row.link_id as string;

					t.links.push({
						created_at: linkCreatedAt,
						updated_at: linkUpdatedAt,
						id: linkId,
						transaction: {
							id: linkedTransactionId,
							date: lDate,
							amount: lAmount,
							currency: lCurrency,
							additional: lAdditional,
							counter_party: lCounterParty,
						},
					});
				}

				transactions.set(id, t);
			}

			const list_transactions = [...transactions.values()];

			const lastId = list_transactions.at(-1)?.id;
			const firstId = list_transactions.at(0)?.id;

			let nextId = null;
			let prevId = null;

			if (hasMore && !cursor) {
				nextId = lastId;
			} else if (cursor?.dir === "left" || (hasMore && cursor?.dir === "right")) {
				nextId = lastId;
			}

			if (cursor?.dir === "right" || (hasMore && cursor?.dir === "left")) {
				prevId = firstId;
			}

			return {
				transactions: list_transactions,
				next_id: nextId,
				prev_id: prevId,
			};
		},

		insertWithCategory: async (props: {
			id: string;
			userId: string;
			date: Date;
			amount: number;
			currency: string;
			counterParty: string;
			additional: string | null;
			categoryName: string;
		}) => {
			const categoryId = id("transaction_category");
			await sql`
				with category as (
					insert into transaction_categories (id, name, user_id)
					values (${categoryId}, ${props.categoryName}, ${props.userId})
					on conflict (user_id, lower(name))
					do update set name = excluded.name
					returning id
				)
				insert into transactions (id, date, amount, currency, counter_party, additional, user_id, category_id)
				values (
					${props.id}, 
					${props.date}, 
					${props.amount}, 
					${props.currency}, 
					${props.counterParty}, 
					${props.additional}, 
					${props.userId}, 
					coalesce((select id from category), ${categoryId})
				)
				returning id;
			`;
		},

		insert: async (props: {
			id: string;
			date: Date;
			amount: number;
			currency: string;
			counterParty: string;
			additional: string | null;
			userId: string;
		}) => {
			await sql`
				insert into transactions (id, date, amount, currency, counter_party, additional, user_id)
				values (${props.id}, ${props.date}, ${props.amount}, ${props.currency}, ${props.counterParty}, ${props.additional}, ${props.userId});
			`;
		},

		insertMany: async (
			transactions: Array<{
				id: string;
				date: Date;
				amount: number;
				currency: string;
				counter_party: string;
				additional: string | null;
				user_id: string;
				category_name?: string;
			}>
		) => {
			const uniqueCategories = new Set<string>();
			for (const t of transactions) {
				if (t.category_name) uniqueCategories.add(t.category_name);
			}

			const categoryIds = new Map<string, string>();

			for (const category of uniqueCategories.values()) {
				const c = {
					id: id("transaction_category"),
					name: category,
					user_id: transactions[0].user_id,
				};

				const [{ id: categoryId }]: [{ id: string }] = await sql`
					insert into transaction_categories (id, name, user_id)
					values (${c.id}, ${c.name}, ${c.user_id})
					on conflict (user_id, lower(name))
					do update set name = excluded.name
					returning id;
				`;

				categoryIds.set(category, categoryId);
			}

			const mapped = transactions.map(({ category_name, ...t }) => {
				const category_id = categoryIds.get(category_name ?? "");

				return {
					...t,
					category_id,
				};
			});

			await sql`insert into transactions ${sql(mapped)}`;
		},

		updateWithCategory: async (props: {
			id: string;
			date: Date;
			amount: number;
			currency: string;
			counterParty: string;
			additional: string | null;
			categoryName: string;
			userId: string;
		}) => {
			const categoryId = id("transaction_category");
			await sql`
				with category as (
					insert into transaction_categories (id, name, user_id)
					values (${categoryId}, ${props.categoryName}, ${props.userId})
					on conflict (user_id, lower(name))
					do update set name = excluded.name
					returning id
				)
				update transactions
				set
					date = ${props.date},
					amount = ${props.amount},
					currency = ${props.currency},
					counter_party = ${props.counterParty},
					additional = ${props.additional},
					category_id = coalesce((select id from category), ${categoryId}),
					updated_at = now()
				where id = ${props.id}
				and user_id = ${props.userId};
			`;
		},

		update: async (props: {
			id: string;
			date: Date;
			amount: number;
			currency: string;
			counterParty: string;
			additional: string | null;
			userId: string;
		}) => {
			await sql`
				update transactions
				set
					date = ${props.date},
					amount = ${props.amount},
					currency = ${props.currency},
					counter_party = ${props.counterParty},
					additional = ${props.additional},
					updated_at = now()
				where id = ${props.id}
				and user_id = ${props.userId};
			`;
		},

		async delete(props: { id: string; userId: string }) {
			await sql`
				delete from transactions
				where id = ${props.id}
				and user_id = ${props.userId};
			`;
		},

		link: async (props: { user_id: string; link_id: string; a_id: string; b_id: string }) => {
			await sql`
				insert into transactions_links (id, transaction_a_id, transaction_b_id, user_id)
				values (${props.link_id}, ${props.a_id}, ${props.b_id}, ${props.user_id})
			`;
		},

		unlink: async (props: { user_id: string; link_id: string }) => {
			await sql`
				delete from transactions_links
				where
					user_id = ${props.user_id}
					and id = ${props.link_id}
			`;
		},
	};
}

export type Transaction = {
	id: string;
	date: Date;
	amount: number;
	currency: string;
	additional: string;
	counter_party: string;
};

export type TransactionLink = {
	id: string;
	created_at: Date;
	updated_at: Date | null;
	transaction: Transaction;
};

export type TransactionWithLinks = Transaction & {
	category: Category | null;
	links: TransactionLink[];
};

export type Category = {
	id: string;
	name: string;
	is_neutral: boolean;
};

export type CategoryWithTxCount = Category & {
	transaction_count: bigint;
};

File: back2/src/data/postgres/transactions.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use futures::TryStreamExt;
use serde::Serialize;
use sqlx::{Postgres, QueryBuilder, prelude::FromRow, query, query_as};
use utoipa::ToSchema;

use crate::data::create_id;

use super::Pool;

#[derive(Clone)]
pub struct Transactions {
    pool: Pool,
}

impl Transactions {
    pub(crate) fn new(pool: Pool) -> Self {
        return Self { pool };
    }

    pub async fn query(&self, user_id: &str) -> Result<Vec<QueryTx>, sqlx::Error> {
        let rows = query_as!(
            QueryTxRow,
            r#"
            select
                t.id as id,
                t.date as date,
                t.counter_party as counter_party,
                t.amount as amount,
                t.category_id as category_id,
                t.currency as currency,
                t.additional as additional,

                c.name as "cat_name?",
                c.is_neutral as "cat_is_ne?",

                link.created_at as "link_created_at?",
                link.updated_at as "link_updated_at?",

                linked.id as "l_id?",
                linked.amount as "l_amount?",
                linked.date as "l_date?",
                linked.counter_party as "l_counter_party?",
                linked.additional as "l_additional?",
                linked.currency as "l_currency?",
                linked.category_id as "l_category_id?"
            from transactions t

            left join transaction_categories c on t.category_id = c.id

            left join transactions_links link
              on link.transaction_a_id = t.id or link.transaction_b_id = t.id
            left join transactions linked
              on (linked.id = CASE WHEN link.transaction_a_id = t.id THEN link.transaction_b_id ELSE link.transaction_a_id END)

            where t.user_id = $1
            "#,
            user_id
        )
        .fetch_all(&self.pool)
        .await?;

        let mut tx_map: HashMap<String, QueryTx> = HashMap::default();

        for row in rows {
            let tx = tx_map.get_mut(&row.id);

            if let Some(tx) = tx {
                if let Some(l_id) = row.l_id {
                    tx.links.push(Link {
                        created_at: row.link_created_at.expect("checked created_at"),
                        updated_at: row.link_updated_at.expect("checked updated_at"),
                        tx: LinkedTx {
                            id: l_id,
                            date: row.l_date.expect("checked linked_date"),
                            counter_party: row
                                .l_counter_party
                                .expect("checked linked_counter_party"),
                            additional: row.l_additional,
                            currency: row.l_currency.expect("checked linked_currency"),
                            amount: row.l_amount.expect("checked linked_amount"),
                        },
                    });
                }
            } else {
                tx_map.insert(
                    row.id.to_owned(),
                    QueryTx {
                        id: row.id,
                        date: row.date,
                        counter_party: row.counter_party,
                        amount: row.amount,
                        additional: row.additional,
                        currency: row.currency,
                        links: vec![],
                        category: if let Some(cat_id) = row.category_id {
                            Some(Category {
                                id: cat_id,
                                name: row.cat_name.expect("checked cat_name"),
                                is_neutral: row.cat_is_ne.expect("checked is_ne"),
                            })
                        } else {
                            None
                        },
                    },
                );
            }
        }

        Ok(tx_map.into_values().collect())
    }

    pub async fn get_by_account_for_sync(
        &self,
        user_id: &str,
        account_id: &str,
    ) -> Result<Vec<SyncTx>, sqlx::Error> {
        let rows = query_as!(
            SyncTx,
            r#"
            select
                t.date,
                t.og_counter_party,
                t.amount as amount
            from transactions t
            where t.user_id = $1 and t.account_id = $2
            "#,
            user_id,
            account_id
        )
        .fetch_all(&self.pool)
        .await?;

        Ok(rows)
    }

    pub async fn insert_with_category(
        &self,
        user_id: &str,
        tx: &InsertTx,
        category_name: &str,
    ) -> Result<(), sqlx::Error> {
        let category_id = create_id();
        let now = Utc::now();
        let updated_at: Option<DateTime<Utc>> = None;

        query!(
            r#"
            with category_id as (
                insert into transaction_categories (
                    user_id,
                    id,
                    created_at,
                    updated_at,
                    name,
                    is_neutral
                )
                values ($1, $2, $3, $4, $5, false)
                on conflict (user_id, lower(name))
                do update set name = excluded.name
                returning id
            )
            insert into transactions (
                user_id,
                id,
                created_at,
                updated_at,
                date,
                amount,
                currency,
                counter_party,
                og_counter_party,
                additional,
                category_id,
                account_id
            )
            values ($1, $6, $3, $4, $7, $8, $9, $10, $10, $11, coalesce((select id from category_id), $2), $12)
            "#,
            user_id,
            category_id,
            now,
            updated_at,
            category_name,
            tx.id,
            tx.date,
            tx.amount,
            tx.currency,
            tx.counter_party,
            tx.additional,
            tx.account_id,
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn insert(&self, user_id: &str, tx: &InsertTx) -> Result<(), sqlx::Error> {
        let now = Utc::now();
        let updated_at: Option<DateTime<Utc>> = None;

        query!(
            r#"
            insert into transactions (
                user_id,
                id,
                created_at,
                updated_at,
                date,
                amount,
                currency,
                counter_party,
                og_counter_party,
                additional,
                account_id
            )
            values ($1, $2, $3, $4, $5, $6, $7, $8, $8, $9, $10)
            "#,
            user_id,
            tx.id,
            now,
            updated_at,
            tx.date,
            tx.amount,
            tx.currency,
            tx.counter_party,
            tx.additional,
            tx.account_id,
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn insert_many(
        &self,
        user_id: &str,
        transactions: Vec<InsertTx>,
    ) -> Result<(), sqlx::Error> {
        let now = Utc::now();
        let updated_at: Option<DateTime<Utc>> = None;

        let mut builder: QueryBuilder<Postgres> = QueryBuilder::new(
            r#"
                insert into transactions (
                    user_id,
                    id,
                    created_at,
                    updated_at,
                    date,
                    amount,
                    currency,
                    counter_party,
                    og_counter_party,
                    additional
                ) 
            "#,
        );

        builder.push_values(transactions, |mut b, tx| {
            b.push_bind(user_id);
            b.push_bind(tx.id);
            b.push_bind(now);
            b.push_bind(updated_at);
            b.push_bind(tx.date);
            b.push_bind(tx.amount);
            b.push_bind(tx.currency);
            b.push_bind(tx.counter_party.to_owned());
            b.push_bind(tx.counter_party);
            b.push_bind(tx.additional);
        });

        let query = builder.build();
        query.execute(&self.pool).await?;

        Ok(())
    }

    pub async fn update<'a>(
        &self,
        user_id: &str,
        tx_id: &str,
        tx: &UpdateTx<'a>,
    ) -> Result<(), sqlx::Error> {
        query!(
            r#"
            update transactions
            set
                updated_at = $3,
                date = $4,
                amount = $5,
                currency = $6,
                counter_party = $7,
                additional = $8
            where user_id = $1 and id = $2
            "#,
            user_id,
            tx_id,
            Utc::now(),
            tx.date,
            tx.amount,
            tx.currency,
            tx.counter_party,
            tx.additional
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn update_with_category<'a>(
        &self,
        user_id: &str,
        tx_id: &str,
        tx: &UpdateTx<'a>,
        category_name: &str,
    ) -> Result<(), sqlx::Error> {
        let category_id = create_id();

        query!(
            r#"
            with category_id as (
                insert into transaction_categories (
                    user_id,
                    created_at,
                    updated_at,
                    name,
                    is_neutral
                )
                values ($1, $2, null, $3, false)
                on conflict (user_id, lower(name))
                do update set name = excluded.name
                returning id
            )
            update transactions
            set
                updated_at = $2,
                date = $5,
                amount = $6,
                currency = $7,
                counter_party = $8,
                additional = $9,
                category_id = coalesce((select id from category_id), $4)
            where user_id = $1 and id = $10
            "#,
            user_id,
            Utc::now(),
            category_name,
            category_id,
            tx.date,
            tx.amount,
            tx.currency,
            tx.counter_party,
            tx.additional,
            tx_id
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn delete(&self, user_id: &str, tx_id: &str) -> Result<(), sqlx::Error> {
        query!(
            r#"
            delete from transactions
            where user_id = $1 and id = $2
            "#,
            user_id,
            tx_id
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn stats(
        &self,
        user_id: &str,
        timezone: &str,
        start: &DateTime<Utc>,
        end: &DateTime<Utc>,
    ) -> Result<HashMap<String, Tx>, sqlx::Error> {
        let mut rows = query_as!(
            TxRow,
            r#"
            select
                t.id as id,
                t.date as date,
                t.counter_party as counter_party,
                t.amount as amount,
                t.category_id as category_id,
                t.additional as additional,
                t.currency as currency,
                c.name as cat_name,
                c.is_neutral as cat_is_ne,

                linked.id as linked_id,
                linked.amount as linked_amount
            from transactions t
            left join transaction_categories c on t.category_id = c.id

            left join transactions_links link on link.transaction_a_id = t.id or link.transaction_b_id = t.id
            left join transactions linked on (
                link.transaction_b_id = linked.id and link.transaction_a_id = t.id
            ) or (
                link.transaction_a_id = linked.id and link.transaction_b_id = t.id
            )

            -- left join transactions_links link
            --   on link.transaction_a_id = t.id or link.transaction_b_id = t.id
            -- left join transactions linked
            --   on (linked.id = CASE WHEN link.transaction_a_id = t.id THEN link.transaction_b_id ELSE link.transaction_a_id END)

            where t.user_id = $1
            and t.date at time zone $2 between $3 and $4;
            "#,
            user_id,
            timezone,
            start.naive_utc(),
            end.naive_utc()
        )
        .fetch(&self.pool);

        let mut tx_map: HashMap<String, Tx> = HashMap::default();

        while let Some(row) = rows.try_next().await? {
            let tx = tx_map.get_mut(&row.id);

            if let Some(tx) = tx {
                if let Some(linked_id) = row.linked_id {
                    tx.links.push(linked_id);
                }
            } else {
                tx_map.insert(
                    row.id.to_owned(),
                    Tx {
                        id: row.id,
                        date: row.date,
                        counter_party: row.counter_party,
                        amount: row.amount,
                        additional: row.additional,
                        currency: row.currency,
                        links: vec![],
                        category: if let Some(cat_id) = row.category_id {
                            Some(Category {
                                id: cat_id,
                                name: row.cat_name.expect("checked cat_name"),
                                is_neutral: row.cat_is_ne.expect("checked is_ne"),
                            })
                        } else {
                            None
                        },
                    },
                );
            }
        }

        return Ok(tx_map);
    }

    pub async fn link(
        &self,
        user_id: &str,
        tx_a_id: &str,
        tx_b_id: &str,
    ) -> Result<(), sqlx::Error> {
        let now = Utc::now();
        let updated_at: Option<DateTime<Utc>> = None;

        query!(
            r#"
            with existing as (
                select user_id from transactions_links
                where user_id = $1
                and (
                    (transaction_a_id = $2 and transaction_b_id = $3)
                    or (transaction_a_id = $3 and transaction_b_id = $2)
                )
            )
            insert into transactions_links (
                user_id,
                created_at,
                updated_at,
                transaction_a_id,
                transaction_b_id
            )
            select $1, $4, $5, $2, $3
            where not exists (select 1 from existing);
            "#,
            user_id,
            tx_a_id,
            tx_b_id,
            now,
            updated_at
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn unlink(
        &self,
        user_id: &str,
        tx1_id: &str,
        tx2_id: &str,
    ) -> Result<(), sqlx::Error> {
        query!(
            r#"
            delete from transactions_links
            where user_id = $1
            and (
                (transaction_a_id = $2 and transaction_b_id = $3)
                or (transaction_a_id = $3 and transaction_b_id = $2)
            )
            "#,
            user_id,
            tx1_id,
            tx2_id,
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }
}

#[derive(Debug, FromRow)]
struct TxRow {
    id: String,
    date: DateTime<Utc>,
    amount: f32,
    counter_party: String,
    additional: Option<String>,
    currency: String,
    category_id: Option<String>,
    cat_name: Option<String>,
    cat_is_ne: Option<bool>,

    linked_id: Option<String>,
    linked_amount: Option<f32>,
}

#[derive(Debug, FromRow)]
struct QueryTxRow {
    id: String,
    date: DateTime<Utc>,
    amount: f32,
    counter_party: String,
    additional: Option<String>,
    currency: String,
    category_id: Option<String>,

    cat_name: Option<String>,
    cat_is_ne: Option<bool>,

    link_created_at: Option<DateTime<Utc>>,
    link_updated_at: Option<DateTime<Utc>>,

    l_id: Option<String>,
    l_date: Option<DateTime<Utc>>,
    l_amount: Option<f32>,
    l_counter_party: Option<String>,
    l_additional: Option<String>,
    l_currency: Option<String>,
    l_category_id: Option<String>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct QueryTx {
    pub id: String,
    pub date: DateTime<Utc>,
    pub counter_party: String,
    pub additional: Option<String>,
    pub currency: String,
    pub category: Option<Category>,
    pub amount: f32,
    pub links: Vec<Link>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct SyncTx {
    pub og_counter_party: String,
    pub date: DateTime<Utc>,
    pub amount: f32,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct Link {
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub tx: LinkedTx,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct LinkedTx {
    pub id: String,
    pub date: DateTime<Utc>,
    pub counter_party: String,
    pub additional: Option<String>,
    pub currency: String,
    pub amount: f32,
}

#[derive(Debug, Clone)]
pub struct Tx {
    pub id: String,
    pub date: DateTime<Utc>,
    pub counter_party: String,
    pub additional: Option<String>,
    pub currency: String,
    pub category: Option<Category>,
    pub amount: f32,
    pub links: Vec<String>,
}

#[derive(Debug)]
pub struct InsertTx {
    pub id: String,
    pub date: DateTime<Utc>,
    pub counter_party: String,
    pub additional: Option<String>,
    pub currency: String,
    pub amount: f32,
    pub account_id: String,
}

#[derive(Debug)]
pub struct UpdateTx<'a> {
    pub date: DateTime<Utc>,
    pub counter_party: &'a str,
    pub additional: Option<&'a str>,
    pub currency: &'a str,
    pub amount: f32,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct Category {
    pub id: String,
    pub name: String,
    pub is_neutral: bool,
}
